src/Agent.Worker/TfManager.cs

```diff
@@ -0,0 +1,131 @@
+
+                if (!File.Exists(downloadTfPath))
+                {
+                    await DownloadAsync(downloadTfPath, tfDownloadUrl, executionContext);
+                }
+                else
+                {
+                    executionContext.Debug($"tf is already downloaded to {downloadTfPath}.");
+                }
+
+                try
+                {
+                    ZipFile.ExtractToDirectory(downloadTfPath, tfLegacyExternalsPath);
+                    File.WriteAllText(downloadTfPath, DateTime.UtcNow.ToString());
+                    executionContext.Debug("tf-legacy has been extracted and cleaned up");
+                }
+                catch (Exception ex)
+                {
+                    executionContext.Error(ex);
+                }
+            }
+            else
+            {
+                executionContext.Debug($"tf-legacy download already exists at {tfLegacyExternalsPath}.");
+            }
+
+            string vstsomLegacyExternalsPath = Path.Combine(externalsPath, "vstsom-legacy");
+
+            if (!Directory.Exists(vstsomLegacyExternalsPath))
+            {
+                const string vstsomDownloadUrl = "https://vstsagenttools.blob.core.windows.net/tools/vstsom/m122_887c6659/vstsom.zip";
+
+                string tempVstsomDirectory = Path.Combine(externalsPath, "vstsom_download_temp");
+                Directory.CreateDirectory(tempVstsomDirectory);
+                string downloadVstsomPath = Path.ChangeExtension(Path.Combine(tempVstsomDirectory, "vstsomdownload"), ".completed");
+
+                if (!File.Exists(downloadVstsomPath))
+                {
+                    await DownloadAsync(downloadVstsomPath, vstsomDownloadUrl, executionContext);
+                }
+                else
+                {
+                    executionContext.Debug($"vstsom is already downloaded to {downloadVstsomPath}.");
+                }
+
+                try
+                {
+                    ZipFile.ExtractToDirectory(downloadVstsomPath, vstsomLegacyExternalsPath);
+                    File.WriteAllText(downloadVstsomPath, DateTime.UtcNow.ToString());
+                    executionContext.Debug("vstsom-legacy has been extracted and cleaned up");
+                }
+                catch (Exception ex)
+                {
+                    executionContext.Error(ex);
+                }
+            }
+            else
+            {
+                executionContext.Debug($"vstsom-legacy download already exists at {vstsomLegacyExternalsPath}.");
+            }
+        }
+
+        private static async Task DownloadAsync(string downloadPath, string blobUrl, IExecutionContext executionContext)
+        {
+            int retryCount = 0;
+            const int retryLimit = 3;
+            const int timeout = 180;
+            const int defaultFileStreamBufferSize = 4096;
+            const int retryDelay = 10000;
+
+            while (retryCount < retryLimit)
```

Comment:
Should we log if the tool wasn't downloaded after the given number of attempts?

Reply:
Added logging for this case